packages/medusa/src/services/payment-provider.ts

```diff
@@ -0,0 +1,544 @@
+      const session = await this.retrieveSession(paymentSession.id)
+      const provider = this.retrieveProvider(paymentSession.provider_id)
+      await provider.withTransaction(transactionManager).deletePayment(session)
+
+      const sessionRepo = transactionManager.getCustomRepository(
+        this.paymentSessionRepository_
+      )
+
+      await sessionRepo.remove(session)
+
+      const sessionData = await provider
+        .withTransaction(transactionManager)
+        .createPayment(cart)
+
+      const toCreate = {
+        cart_id: cart.id,
+        provider_id: session.provider_id,
+        data: sessionData,
+        is_selected: true,
+        status: "pending",
+      }
+
+      const created = sessionRepo.create(toCreate)
+      return await sessionRepo.save(created)
+    })
+  }
+
+  /**
+   * Updates an existing payment session.
+   * @param paymentSession - the payment session object to
+   *    update
+   * @param cart - the cart object to update for
+   * @return the updated payment session
+   */
+  async updateSession(
+    paymentSession: PaymentSession,
+    cart: Cart
+  ): Promise<PaymentSession> {
+    return await this.atomicPhase_(async (transactionManager) => {
+      const session = await this.retrieveSession(paymentSession.id)
+      const provider = this.retrieveProvider(paymentSession.provider_id)
+      session.data = await provider
+        .withTransaction(transactionManager)
+        .updatePayment(paymentSession.data, cart)
+
+      const sessionRepo = transactionManager.getCustomRepository(
+        this.paymentSessionRepository_
+      )
+      return sessionRepo.save(session)
+    })
+  }
+
+  async deleteSession(
+    paymentSession: PaymentSession
+  ): Promise<PaymentSession | undefined> {
+    return await this.atomicPhase_(async (transactionManager) => {
+      const session = await this.retrieveSession(paymentSession.id).catch(
+        () => void 0
+      )
+
+      if (!session) {
+        return
+      }
+
+      const provider = this.retrieveProvider(paymentSession.provider_id)
+      await provider
+        .withTransaction(transactionManager)
+        .deletePayment(paymentSession)
+
+      const sessionRepo = transactionManager.getCustomRepository(
+        this.paymentSessionRepository_
+      )
+
+      return sessionRepo.remove(session)
+    })
+  }
+
+  /**
+   * Finds a provider given an id
+   * @param {string} providerId - the id of the provider to get
+   * @return {PaymentService} the payment provider
+   */
+  retrieveProvider<
+    TProvider extends AbstractPaymentService<never> | typeof BasePaymentService
+  >(
+    providerId: string
+  ): TProvider extends AbstractPaymentService<never>
+    ? AbstractPaymentService<never>
+    : typeof BasePaymentService {
+    try {
+      let provider
+      if (providerId === "system") {
+        provider = this.container_[`systemPaymentProviderService`]
+      } else {
+        provider = this.container_[`pp_${providerId}`]
```

Comment:
thought: Should we follow same pattern as with the notification provider i.e. defining a type `PaymentProviderKey`?